datafusion-federation/src/optimize/optimize_projections/mod.rs

```diff
@@ -0,0 +1,1047 @@
+    indices: RequiredIndicies,
+    join_type: &JoinType,
+) -> (RequiredIndicies, RequiredIndicies) {
+    match join_type {
+        // In these cases requirements are split between left/right children:
+        JoinType::Inner
+        | JoinType::Left
+        | JoinType::Right
+        | JoinType::Full
+        | JoinType::LeftMark => {
+            // Decrease right side indices by `left_len` so that they point to valid
+            // positions within the right child:
+            indices.split_off(left_len)
+        }
+        // All requirements can be re-routed to left child directly.
+        JoinType::LeftAnti | JoinType::LeftSemi => (indices, RequiredIndicies::new()),
+        // All requirements can be re-routed to right side directly.
+        // No need to change index, join schema is right child schema.
+        JoinType::RightSemi | JoinType::RightAnti => (RequiredIndicies::new(), indices),
+    }
+}
+
+fn add_projection_on_top_if_helpful(
+    plan: LogicalPlan,
+    _project_exprs: Vec<Expr>,
+    _config: &dyn OptimizerConfig,
+) -> Result<Transformed<LogicalPlan>> {
+    // Always prefer the original query layout.
+    Ok(Transformed::no(plan))
+}
+
+/// Rewrite the given projection according to the fields required by its
+/// ancestors.
+///
+/// # Parameters
+///
+/// * `proj` - A reference to the original projection to rewrite.
+/// * `config` - A reference to the optimizer configuration.
+/// * `indices` - A slice of indices representing the columns required by the
+///   ancestors of the given projection.
+///
+/// # Returns
+///
+/// A `Result` object with the following semantics:
+///
+/// - `Ok(Some(LogicalPlan))`: Contains the rewritten projection
+/// - `Ok(None)`: No rewrite necessary.
+/// - `Err(error)`: An error occured during the function call.
+fn rewrite_projection_given_requirements(
+    proj: Projection,
+    config: &dyn OptimizerConfig,
+    indices: &RequiredIndicies,
+) -> Result<Transformed<LogicalPlan>> {
+    let Projection { expr, input, .. } = proj;
+
+    let exprs_used = indices.get_at_indices(&expr);
+
+    let required_indices = RequiredIndicies::new().with_exprs(input.schema(), exprs_used.iter());
+
+    // rewrite the children projection, and if they are changed rewrite the
+    // projection down
+    optimize_projections(Arc::unwrap_or_clone(input), config, required_indices)?.transform_data(
+        |input| {
+            if is_projection_unnecessary(&input, &exprs_used, config)? {
+                Ok(Transformed::yes(input))
+            } else {
+                Projection::try_new(exprs_used, Arc::new(input))
+                    .map(LogicalPlan::Projection)
+                    .map(Transformed::yes)
+            }
+        },
+    )
+}
+
+fn is_projection_unnecessary(
+    _input: &LogicalPlan,
+    _proj_exprs: &[Expr],
+    _config: &dyn OptimizerConfig,
+) -> Result<bool> {
+    // Always prefer the original query layout. Return false to keep the projection.
+    Ok(false)
```

Comment:
Alternatively we can remove usage of `is_projection_unnecessary` completely but this approach simplifies future updates as we modify few helper functions only w/o changing core logic => easy to upgrade moving forward (if it will be required)